src/company_cmd.cpp

```diff
@@ -980,15 +980,37 @@ CommandCost CmdCompanyCtrl(DoCommandFlag flags, CompanyCtrlAction cca, CompanyID
 }
 
 /**
- * Add the given public key to the allow list of this company.
+ * Add or remove the given public key to the allow list of this company.
  * @param flags Operation to perform.
- * @param public_key The public key of the client to add.
+ * @param action The action to perform.
+ * @param public_key The public key of the client to add or remove.
  * @return The cost of this operation or an error.
  */
-CommandCost CmdCompanyAddAllowList(DoCommandFlag flags, const std::string &public_key)
+CommandCost CmdCompanyAllowListCtrl(DoCommandFlag flags, CompanyAllowListCtrlAction action, const std::string &public_key)
 {
+	Company *c = Company::GetIfValid(_current_company);
+	if (c == nullptr) return CMD_ERROR;
+
+	/* The public key length includes the '\0'. */
+	if (public_key.size() != NETWORK_PUBLIC_KEY_LENGTH - 1) return CMD_ERROR;
+
+	using ModifierFunction = bool (NetworkAuthorizedKeys::*)(std::string_view key);
+	ModifierFunction modifier_function;
+	switch (action) {
+		case CALCA_ADD:
+			modifier_function = &NetworkAuthorizedKeys::Add;
+			break;
+
+		case CALCA_REMOVE:
+			modifier_function = &NetworkAuthorizedKeys::Remove;
+			break;
+
+		default:
+			return CMD_ERROR;
+	}
+
 	if (flags & DC_EXEC) {
-		if (Company::Get(_current_company)->allow_list.Add(public_key)) {
+		if (std::invoke(modifier_function, c->allow_list, public_key)) {
```

Comment:
Seems quite a complex way to choose whether to Add or Remove, are more operations planned?

Reply:
I have not planned any yet.

It would mean two switch statements; one for validation and one for actually executing the command.